dtm0/service.c

```diff
@@ -233,33 +234,50 @@ m0_dtm0_service_process_connect(struct m0_reqh_service *s,
 	return M0_RC(rc);
 }
 
-M0_INTERNAL int
-m0_dtm0_service_process_disconnect(struct m0_reqh_service *s,
-				   struct m0_fid          *remote_srv)
+static int dtm0_process_disconnect(struct dtm0_process *process)
 {
-	int rc;
-	struct dtm0_process *process = NULL;
+	const m0_time_t timeout =
+		m0_time_from_now(DTM0_DISCONNECT_TIMEOUT_SECS, 0);
+	int             rc;
 
-	M0_LOG(M0_DEBUG, "dtm0=%p, remote_srv="FID_F, s, FID_P(remote_srv));
+	M0_ENTRY("process=%p, rfid=" FID_F, process,
+		 FID_P(&process->dop_rserv_fid));
 
-	process = dtm0_service_process__lookup(s, remote_srv);
-	if (process == NULL)
-		return M0_RC(-ENOENT);
+	if (M0_IS0(&process->dop_rlink))
+		return M0_RC(0);
 
-	rc = m0_rpc_link_disconnect_sync(
-		&process->dop_rlink,
-		m0_time_from_now(DTM0_DISCONNECT_TIMEOUT_SECS, 0));
+	rc = m0_rpc_link_is_connected(&process->dop_rlink) ?
+		m0_rpc_link_disconnect_sync(&process->dop_rlink, timeout) : 0;
 
 	if (rc == -ETIMEDOUT) {
-		M0_LOG(M0_WARN, "Disconnect timeout");
+		M0_LOG(M0_WARN, "Disconnect timeout (suppressed)");
 		rc = 0;
 	}
 
+	if (rc != 0)
+		return M0_ERR(rc);
+
 	m0_rpc_link_fini(&process->dop_rlink);
+	M0_SET0(&process->dop_rlink);
 
 	return M0_RC(rc);
 }
 
+M0_INTERNAL int
+m0_dtm0_service_process_disconnect(struct m0_reqh_service *s,
+				   struct m0_fid          *remote_srv)
+{
+	struct dtm0_process *process;
```

Comment:
nit: `struct dtm0_process *process = dtm0_service_process__lookup(s, remote_srv);`

Reply:
done, and I also folded the return statements.